packages/babel-parser/src/plugins/jsx/index.js

```diff
@@ -559,7 +559,9 @@ export default (superClass: Class<Parser>): Class<Parser> =>
       if (
         code === charCodes.lessThan &&
         this.state.exprAllowed &&
-        this.input.charCodeAt(this.state.pos + 1) !== charCodes.exclamationMark
+        this.input.charCodeAt(this.state.pos + 1) !==
+          charCodes.exclamationMark &&
+        context !== tc.functionStatement
```

Comment:
Also `functionExpression`.
However, I think that it would be better to set `this.state.exprAllowed` to `false` when we find `of` after `function`, rather than working it around here.

Reply:
I'll check this.